### DISCUSSION TITLE
使用体验 / Feedback

### DISCUSSION THREAD

**Discussion by AkutaZehy** [otomad/om_midi#16] [Show and tell]
这两天试用了一下，总体来说还是挺好用的吧，引入多轨，生成的轨道是带命名和根据midi截断的这一点挺让我印象深刻的，然后下面来点建议。<br>
Recently used this script, pretty in all. Multiplied channels. named channel and just part of the track according to midi are really impressive add-ons. Followings are suggestions.

![image](https://user-images.githubusercontent.com/42381419/196014706-e9654977-e37d-4362-8302-350902625298.png)
看密恐了 / Trypophobia lmao
<hr>

> 面板实在是太大了
> Too large the panel.

这个可以理解吧，毕竟功能确实是太多了，如果要手动加/删一些功能的话应该咋调？抱着好奇心开了一下发现代码跟毛线球一样...不知道用AE脚本专用的那个编辑器会不会好点。然后有些比较怪的子面板粘下面了希望看看是bug还是设计问题。<br>
Understanding, so how could I edit the functions and keep those I need? Messy codes really disturb...If the special editor for AE Script could be better. And weird sub panels btw.

<center>
<img src="https://user-images.githubusercontent.com/42381419/196014296-78602fc5-9374-4566-8d80-ceaac68e8c46.png" >
每次都要手动拉开 / Pull to full size every time<br>
</center>

<hr>

> 界面问题
> Panel Problems

<center>
<img src="https://user-images.githubusercontent.com/42381419/196014254-4e80c541-1284-4847-9af2-9a24f9247bae.png">
一个长度很奇怪的下拉菜单 / A menu with weird size
</center>

<center>
<img src="https://user-images.githubusercontent.com/42381419/196014558-bbb6d047-ef2b-46db-b0c1-fd76ceee4522.png">
我还是更希望它的“应用”按钮能放在上面 / Wish the Apply button could be put on the top area

<hr>

## 一些小小的疑问 / Questions

可能是专栏没提到~~或者是我看漏了~~，一些比较好奇的点就单独扔这问了。
Not mentioned ~~or I just overlooked them~~, some curious points.

1. 关于变速 / About Varying BPM
因为看到有个专门的BPM面板，比较好奇变速（即BPM通道包络）曲子的帧生成会不会异常（还没试过），比如严重的帧对不上的情况。
Noticed an additional BPM input, I've never tried but curious about: will a midi file with BPM envelope be generated as a track comes up with problems such as fatal error on frames alignment.

2. 关于非对齐帧 / About generated frames' offset
众所周知，AE的帧数对上曲速其实是一件很困难的事，要用60帧率去做BPM130/170/220或者其他偏的更离谱的是很常见的。在前面的一些版本中，这个偏移会导致更严重的问题，如异常的时间重映射和翻转动画等，不知道这一版改进了这个问题没。（因为我个人是用的表达式时间重映射回避掉了这个问题所以我可能用不到只是顺便问问w）
As we all know, it's hard to let the frames align with the tempo for we always make otomads using 60fps + 130/170/220/etc. BPM. In preceding versions, the offset will lead to fatal errors on Time Remap and Mirroring animations. I wonder if the problem is fixed in this version. (However, I am using Expression-assisted Time Remap, so it never happened to me, just asked btw.)

**Comment by otomad** [comment]
> 面板实在是太大了
> Too large the panel

确实，这一版由于属性加太多了（并且很多都是不常用属性，*我可以不用，但不能没有*）。我个人喜爱将面板固定在 AE 右侧的面板组中，然而我的屏幕高度甚至无法在一页中全部显示出来。我正在想一些能压缩面板的更好的方法。

Indeed, this version has too many properties added (and many are infrequent properties, *I can use without it, but we can't use without it*). I personally like to pin it in the panel group on the right side of AE, however my screen height doesn't even fit all on one page. I'm thinking of some better ways to compress this panel.

> 每次都要手动拉开
> Pull to full size every time

您可以点击此面板右上角三条纹按钮，然后取消勾选 *面板组设置 > 堆叠的面板组*，此时应该会有更好的显示效果。

You can click the three stripes button in the upper right corner of this panel and uncheck *Panel Group Settings > Stacked Panel Groups*, which should have a better display.

![image](https://user-images.githubusercontent.com/56647156/196021150-9ca4d2c6-2174-48aa-9015-fe20ecdba6da.png)

> 代码跟毛线球一样
> Code like wool ball

发行版中的代码是经过压缩后的。如果想看源代码，可以在 GitHub 中查看。所有内容是分模块编写的。

The code in the release is compressed. If you want to see the source code, you can view it in the GitHub. All are written in modules.

> AE 脚本专用的那个编辑器
> That editor for AE scripts

您说的应该是 Adobe ExtendScript Toolkit 吧，那个工具已经很久没有更新了，并且官方也已经不推荐使用了。官方最新推荐使用的编辑器就是 Visual Studio Code。

You should be talking about the Adobe ExtendScript Toolkit, which has not been updated for a long time and is not officially recommended. The official latest recommended editor is Visual Studio Code.

> 一个长度很奇怪的下拉菜单
> A menu with weird size

呃，这个似乎是 Adobe 的锅。如果将面板拉伸得比较宽，下拉菜单就会跟着变得那么宽。如果再将面板拉伸得比较窄，下拉菜单还是刚才那么宽。

Emm, this seems to be Adobe's responsibility. If you resize the panel wider, the drop-down menu becomes wider. Then if you resize the panel narrower, the drop-down menu will still be as wide as it was.

> 我还是更希望它的“应用”按钮能放在上面
> Wish the Apply button could be put on the top area

等我把“面板实在是太大了”这个问题先解决了吧……

Let me solve the "Too large the panel" problem first...

> 关于变速
> About Varying BPM

目前版本（v3.9.10.0）还不支持变速，此时将始终保持为用户所设定的常速。下一版本会支持，[隔壁项目](https://github.com/otomad/OtomadHelper)已经支持了。

The current version (v3.9.10.0) does not support dynamic tempo, which will always keep on the user-set static tempo. The next version will support it, and [the another project](https://github.com/otomad/OtomadHelper) already supports it.

> 关于非对齐帧
> About generated frames' offset

之前的版本和这一版本都是非对齐帧，即不管合成的帧率如何，始终以 BPM 为准。点击生成后的关键帧，会发现它会自动量化为帧（将该关键帧偏移到当前帧率所能表示的最接近的帧上）。一种简单的解决方法是将帧率设置为 BPM 值，然后在渲染前调整回原帧率。至于之前版本中异常的时间重映射和翻转动画等，个人猜测是因为 MIDI 中包含了和弦/复音，这种情况在旧版中会检测不准确，如果是这种情况的话该版是优化了一些的。<s>（其实就是我个人也不太清楚这种情况究竟如何，如果发生了类似问题，以后再来单独解决。）</s>
